metric_name_builder.go

```diff
@@ -81,29 +80,40 @@ var perUnitMap = map[string]string{
 	"y":  "year",
 }
 
-// BuildCompliantMetricName builds a Prometheus-compliant metric name for the specified metric.
-//
-// Metric name is prefixed with specified namespace and underscore (if any).
-// Namespace is not cleaned up. Make sure specified namespace follows Prometheus
-// naming convention.
+// MetricNameBuilder is a helper struct to build metric names.
+type MetricNameBuilder struct {
+	Namespace          string
+	WithMetricSuffixes bool
+	UTF8Allowed        bool
+}
+
+// Build builds a metric name for the specified metric.
 //
+// If UTF8Allowed is true, the metric name is returned as is, only with the addition of type/unit suffixes and namespace preffix if required.
+// Otherwise the metric name is normalized to be Prometheus-compliant.
 // See rules at https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels,
 // https://prometheus.io/docs/practices/naming/#metric-and-label-naming
-// and https://github.com/open-telemetry/opentelemetry-specification/blob/v1.38.0/specification/compatibility/prometheus_and_openmetrics.md#otlp-metric-points-to-prometheus.
-func BuildCompliantMetricName(metric pmetric.Metric, namespace string, addMetricSuffixes bool) string {
+func (b *MetricNameBuilder) Build(name, unit string, metricType MetricType) string {
```

Comment:
How about making a type for the input. I think you can have

```go
// this is not Java - I propose to name it Namer and not Builder :wink: 
// the parameters define how the configuration for the namer - not the input
type MetricNamer struct {
	WithMetricSuffixes bool
	UTF8Allowed        bool
}

type Metric struct {
	Type MetricType
	Name string
	Unit string
	Namespace string
}
```

Reply:
Interesting, indeed, that bulletproofs the API for evolvability, but maybe we're being over precatious? We're talking about translating names between two conventions, and the good thing about conventions is that they don't change over time.

I can't see a reason to add more properties to metrics in the future 🤔

Reply:
And no problems changing the struct to Namer instead of Builder :) No strong opinions on that

Reply:
> good thing about conventions is that they don't change over time

Once there is a new "metric feature" they will need to evolve.
Do you see any downsides of such API?

Reply:
> Do you see any downsides of such API?

Not really, it's just what is more pleasant to my eyes :P

```go
namer := otlptranslator.MetricNamer{}

metricName := namer.Build(m.Name(), m.Unit(), m.Type()) // Smaller, easier to look and understand
metricName := name.Build(otlptranslator.Metric{Name: m.Name(), Unit: m.Unit(), Type: m.Type()})
```

Reply:
Oh wait, curious question since I don't know the answer.

Will we do an extra allocation to create the Metric struct? Do we allocate to store the variable address?

Reply:
> will we do an extra allocation to create the Metric struct? Do we allocate to store the variable address?

Check the Go escape analysis and benchmarks:
- https://goperf.dev/01-common-patterns/stack-alloc/
- https://topofmind.dev/blog/2023/07/15/analyzing-go-heap-escapes/
- https://medium.com/@pranoy1998k/understanding-escape-analysis-in-go-b2db76be58f0
- https://syntactic-sugar.dev/blog/nested-route/go-escape-analysis

I am pretty sure that for most cases the Metric struct will be allocated on the stack (and not do any extra heap allocation).

Reply:
PR updated, [PoC](https://github.com/prometheus/prometheus/pull/16626) as well!

Reply:
Please resolve the comment :wink:

Reply:
Side note: I think you can consider the same API for evolvability and similarity  for normalizing label names

```go
type LabelNamer struct {
}

type Label struct {
	Name string
}
```

Reply:
@ArthurSens, do you think that we should create an issue for https://github.com/prometheus/otlptranslator/pull/29#discussion_r2103320996?